compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -493,7 +493,22 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
     });
 
     // #73631: closures inherit `#[target_feature]` annotations
-    if tcx.features().target_feature_11 && tcx.is_closure(did.to_def_id()) {
+    //
+    // If this closure is marked `#[inline(always)]`, simply skip adding `#[target_feature]`.
+    //
+    // At this point, `unsafe` has already been checked and `#[target_feature]` only affects codegen.
+    // Emitting both `#[inline(always)]` and `#[target_feature]` can potentially result in an
+    // ICE, because LLVM errors when the function fails to be inlined due to a target feature
+    // mismatch.
+    //
+    // Using `#[inline(always)]` implies that this closure will most likely be inlined into
+    // its parent function, which effectively inherits the features anyway. Boxing this closure
+    // would result in this closure being compiled without the inherited target features, but this
+    // is probably a poor usage of `#[inline(always)]` and easily avoided by not using the attribute.
```

Comment:
Boxing seems like a waste, yes, but now that I am thinking about it, this seems like it could result in confusing behavior in the "escaping closure" case, when that would result, instead of the IIFE? Does that even make sense?

Reply:
Inlining with closures is unfortunately always confusing.  `Box<dyn FnOnce()>`, for example, implements `FnOnce` itself: https://github.com/rust-lang/rust/blob/c4083faade32b16feb471ef082ce0b1e3fe94262/library/alloc/src/boxed.rs#L2003-L2009

_This_ `call_once` doesn't have any inline attribute at all! Therefore, the boxed closure's `call_once` inlines into this `call_once`, and then it's up in the air after that.